packages/eslint-plugin/src/rules/no-state-access.ts

```diff
@@ -0,0 +1,112 @@
+import { matchModel } from "../utils/matchModel";
+
+const rule: TSRuleModule = {
+  meta: {
+    docs: {
+      description:
+        "Requires that the state variable is only accessed inside a component or action",
+      category: "Possible Errors",
+      recommended: true,
+    },
+    messages: {
+      accessingState:
+        "Trying to access the state variable. This is empty outside components or functions.",
+    },
+    schema: [],
+    type: "problem",
+  },
+  create(context: TSRuleContext) {
+    const parserServices: ParserServices = tsPluginUtil.getParserServices(
+      context,
+    );
+    const checker = parserServices.program!.getTypeChecker();
+    let importsStateFromModel: boolean = false;
+    let stateImportSymbol: ts.Symbol;
+    const insideFunction: number[] = [];
+    let insideImportDeclaration: boolean = false;
+
+    return {
+      FunctionDeclaration(node: TSESTree.FunctionDeclaration) {
+        const functionTsNode = parserServices.esTreeNodeToTSNodeMap!.get<
+          TSNode
+        >(node);
+        insideFunction.push(functionTsNode.pos);
+      },
+      FunctionExpression(node: TSESTree.FunctionExpression) {
+        const functionTsNode = parserServices.esTreeNodeToTSNodeMap!.get<
+          TSNode
+        >(node);
+        insideFunction.push(functionTsNode.pos);
+      },
+      ArrowFunctionExpression(node: TSESTree.ArrowFunctionExpression) {
+        const functionTsNode = parserServices.esTreeNodeToTSNodeMap!.get<
+          TSNode
+        >(node);
+        insideFunction.push(functionTsNode.pos);
+      },
+      "FunctionDeclaration:exit"() {
+        insideFunction.pop();
+      },
+      "FunctionExpression:exit"() {
+        insideFunction.pop();
+      },
+      "ArrowFunctionExpression:exit"() {
+        insideFunction.pop();
+      },
+
+      ImportDeclaration(node: TSESTree.ImportDeclaration): void {
+        insideImportDeclaration = true;
+        const specifiers = node.specifiers;
+        if (node.source.type === AST_NODE_TYPES.Literal) {
+          if (matchModel(node.source.value as string)) {
+            specifiers.forEach(specifier => {
+              if (specifier.type === AST_NODE_TYPES.ImportSpecifier) {
+                const specifierTsNode = parserServices.esTreeNodeToTSNodeMap!.get<
+                  TSNode
+                >(specifier);
+                if (specifierTsNode.getFullText() === "state") {
+                  importsStateFromModel = true;
+                  stateImportSymbol = (specifierTsNode as any).symbol;
+                }
+              }
+            });
+          }
+        }
+      },
+      "ImportDeclaration:exit"() {
+        insideImportDeclaration = false;
+      },
+
+      Identifier(node: TSESTree.Identifier): void {
+        if (
+          !importsStateFromModel ||
+          insideFunction.length !== 0 ||
+          insideImportDeclaration
+        ) {
+          return;
+        }
```

Comment:
I would prefer to use `node.parent` to e.g. check that we're inside an import declaration.

Reply:
Am I also right in assuming that we don't care about which functions are actions vs just general functions?

Reply:
Oh, and this would throw an error on:

```js
import { state } from "./model";
const myState = foo.state;
// or
const { state: myState } = foo;
```

You need to check that the identifier is actually referencing a variable (rather than a non-computed property etc)

Reply:
Finally, I'm surprised that there isn't a way to get the binding of an identifier - didn't you mention this earlier? 

For example, given:

```js
import { state } from "./model";
const foo = (state) => {
  state.foo = "foo";
};
```

we would want to know that the `state` in line 3 refers to the `ArrowFunctionExpression`'s argument, rather than the `ImportSpecifier`, and therefore not throw a linting exception.

Reply:
Regarding the last point, it is ```checker.getSymbolAtLocation(identifierTsNode);
        if (symbol && symbol === stateImportSymbol)```

Reply:
Why `node.parent` though? Isn't that dirtier? What if we have a `DefaultImportSpecifier` vs a bunch of specifiers, vs something more complicated? More edge cases.

Reply:
Aren't functions that use state actions by definition?

Reply:
No, to avoid nested actions. Actions have to be at the root level.

Reply:
Re `node.parent` - it seems cleaner to me. There aren't that many edge cases, because all specifiers have to be the direct child of an `ImportDeclaration`. It also make the logic much easier to follow:

1. Visit all identifiers.
2. Check whether the identifier is `state` and comes from the model file. If no, pass.
3. Check whether the parent is an import statement. If yes, pass.
4. Check whether we're in an action. If yes, pass.
5. Throw an error.

Btw you can probably use selectors to achieve the same thing. I think:

```
Identifier:not(:matches(:function, ImportDeclaration) Identifier)
```

might work